packages/ifpack2/src/Ifpack2_Experimental_RBILUK_def.hpp

```diff
@@ -1088,42 +1093,61 @@ apply (const Tpetra::MultiVector<scalar_type,local_ordinal_type,global_ordinal_t
-      // Kokkos kernels impl. For now, the only block trsv available is Sequential
-      // and must be done on host.
-      using row_map_type = typename local_matrix_host_type::row_map_type;
-      using index_type   = typename local_matrix_host_type::index_type;
-      using values_type  = typename local_matrix_host_type::values_type;
-
-      auto X_view = X.getLocalViewHost(Tpetra::Access::ReadOnly);
-      auto Y_view = Y.getLocalViewHost(Tpetra::Access::ReadWrite);
-
-      auto L_row_ptrs_host = L_block_->getCrsGraph().getLocalRowPtrsHost();
-      auto L_entries_host = L_block_->getCrsGraph().getLocalIndicesHost();
-      auto U_row_ptrs_host = U_block_->getCrsGraph().getLocalRowPtrsHost();
-      auto U_entries_host = U_block_->getCrsGraph().getLocalIndicesHost();
-      auto L_values_host = L_block_->getValuesHost();
-      auto U_values_host = U_block_->getValuesHost();
-
-      row_map_type* L_row_ptrs_host_ri = reinterpret_cast<row_map_type*>(&L_row_ptrs_host);
-      index_type* L_entries_host_ri    = reinterpret_cast<index_type*>(&L_entries_host);
-      row_map_type* U_row_ptrs_host_ri = reinterpret_cast<row_map_type*>(&U_row_ptrs_host);
-      index_type* U_entries_host_ri    = reinterpret_cast<index_type*>(&U_entries_host);
-      values_type* L_values_host_ri    = reinterpret_cast<values_type*>(&L_values_host);
-      values_type* U_values_host_ri    = reinterpret_cast<values_type*>(&U_values_host);
+      // Kokkos kernels impl.
+      auto X_views = X.getLocalViewDevice(Tpetra::Access::ReadOnly);
+      auto Y_views = Y.getLocalViewDevice(Tpetra::Access::ReadWrite);
+
+      auto lclL = L_block_->getLocalMatrixDevice();
+      auto L_rowmap  = lclL.graph.row_map;
+      auto L_entries = lclL.graph.entries;
+      auto L_values  = lclL.values;
+
+      auto lclU = U_block_->getLocalMatrixDevice();
+      auto U_rowmap  = lclU.graph.row_map;
+      auto U_entries = lclU.graph.entries;
+      auto U_values  = lclU.values;
 
       const auto numRows = L_block_->getLocalNumRows();
-      local_matrix_host_type L_block_local_host("L_block_local_host", numRows, numRows, L_entries_host.size(), *L_values_host_ri, *L_row_ptrs_host_ri, *L_entries_host_ri, blockSize_);
-      local_matrix_host_type U_block_local_host("U_block_local_host", numRows, numRows, U_entries_host.size(), *U_values_host_ri, *U_row_ptrs_host_ri, *U_entries_host_ri, blockSize_);
 
       if (mode == Teuchos::NO_TRANS) {
-        KokkosSparse::trsv("L", "N", "N", L_block_local_host, X_view, Y_view);
-        KokkosSparse::trsv("U", "N", "N", U_block_local_host, Y_view, Y_view);
-        KokkosBlas::axpby(alpha, Y_view, beta, Y_view);
+        KokkosSparse::Experimental::SPTRSVAlgorithm alg = KokkosSparse::Experimental::SPTRSVAlgorithm::SEQLVLSCHD_RP;
+        {
+          KernelHandle_->create_sptrsv_handle(alg, numRows, true /*lower*/, blockSize_);
+          KokkosSparse::Experimental::sptrsv_symbolic(KernelHandle_.getRawPtr(), L_rowmap, L_entries, L_values);
```

Comment:
I didn't notice this on the first review but this shouldn't create a new handle and call sptrsv_symbolic for each apply call. That should happen in ``RBILUK::compute()`` instead. I would say just add two more handles to the RBILUK class, one for L solve and one for U solve. This is what the non block version, ``Ifpack2::RILUK`` does (though each handle is stored in LocalSparseTriangularSolver class).

Reply:
Good idea.

Reply:
Also since these kernels are executing on the same (default) stream, none of the fences in this function are necessary

Reply:
@brian-kelley , thanks! This was a nice catch. I remember thinking that doing the sptrsv handle creation on every apply was only a temporary hack while I debugged sptrsv_numeric, but then I forgot to clean it up. I also forgot to change the algorithm back to TP1.